Adding initial HAWQ support into gptransfer utility

kdunn926 opened the pull request:
PDE occasionally gets a request for a quick way to offload data from GPDB into HAWQ. This PR contains on-the-fly schema fixups and forces HAWQ virtual segments for the INSERT to match the source GPDB system. Here is a sample invocation:

```
gptransfer -d gptest --delimiter='|' 
--source-host=192.168.96.238 --dest-host=192.168.97.94 
--dest-port=10432 
--dest-is-hawq
--source-map-file=host_map_file --format=TEXT --sub-batch-size=2 --batch-size=10
```

`sub-batch-size` should be **at least** the number of GPDB source segment servers and **at most** the total number of GPDB segments.

**--dest-is-hawq** triggers the newly added functionality.

Note, this assumes the target is HAWQ version 2.0.

ivannovick commented:
Great feature!  Thx @kdunn926 

@deepesh1081 can you help review this patch with some folks in your team?  Also interesting for @mkiyama to see this

kaknikhil commented:
@kdunn926 : is there a way to test this patch ? currently gptransfer has a behave test suite (see gptransfer.feature). it would be great to have some test(s) associated with the patch

kdunn926 commented:
@kaknikhil - I'll work on implementing some testing. Some quick and dirty testing was done between GPDB and HDB sandboxes but thorough testing will certainly bubble up some more edge cases.

kaknikhil commented on gpMgmt/bin/gptransfer:
```diff
@@ -1677,6 +1689,35 @@ GROUP BY nspname, relname;''' % (self._table_pair.source.schema,
+                if "CREATE EXTERNAL TABLE" in line:
+                    tableName = line.split(" ")[3]
+                if "LOG ERRORS" in line:
+                    newSql.append(re.sub("LOG ERRORS", "LOG ERRORS INTO {0}_err".format(tableName), line))
```
i am not sure about the logic here but there is a possibility for tableName to be
1. None 
 or
2. a value from the previously executed iteration(since it's only initialized once at 1695) . 

both of these cases may make the sql invalid or incorrect, right ? or is there a reason that it won't ? in any case , adding comments will help a lot.

kdunn926 replied:
For 1) You are correct, for everything except external tables, it *will* be None, which is OK since it is only used when LOG ERRORS is also found (i.e. in External Table DDL). There should probably be checks here - the logic is basically for mapping the DDL for GPDB external tables to equivalent DDL for HAWQ external tables - the latter requires the error table name to be explicitly specified. Given the top level function invoking this logic is called only once per table, it is safe to say that 2) could not happen.

kaknikhil commented:
instead of saying --dest-is-hawq=1 , how do you feel about just saying something on the lines of gptransfer --hawq-as-dest ?? IMHO , i think this makes it more intuitive  . thoughts?

kdunn926 commented:
I agree, --dest-is-hawq is better. I originally tried type='bool' in the argparse routine but that didin't work so I just switched it to int and moved on.

danielgustafsson commented on gpMgmt/bin/gptransfer:
```diff
@@ -512,6 +512,16 @@ def create_parser():
         metavar='<database>',
         help='Database to transfer all source tables into'
     )
+    # destination HAWQ flasg
```
s/flasg/flags/

kdunn926 replied:
Fixed.

kaknikhil commented:
@kdunn-pivotal : any updates ?

kdunn-pivotal commented:
@kaknikhil - Unfortunately I do not have cycles now or in the near future to implement a way to do automated testing for this feature. I do know @zodom-pivotal has field tested this with a customer.

ivannovick commented:
Lets close this as we don't have traction....

kdunn-pivotal commented:
This is a strange reason to leave features on the table.
